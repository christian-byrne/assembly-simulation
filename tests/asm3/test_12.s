 # For this test, we'll use a modified version of getNextLetter from test01
.globl getNextLetter
getNextLetter:
	# standard prologue
	addiu  $sp, $sp, -24
	sw     $fp, 0($sp)
	sw     $ra, 4($sp)
	addiu  $fp, $sp, 20

.data
getNextLetter_BUF:	.asciiz "vaoijwhiewhtowehtweoaiowtt"
			.word   0    # make sure we never overrun the buffer!
			.word   0    # make sure we never overrun the buffer!
			.word   0    # make sure we never overrun the buffer!
			.word   0    # make sure we never overrun the buffer!
			.word   0    # make sure we never overrun the buffer!
			.word   0    # make sure we never overrun the buffer!
			.word   0    # make sure we never overrun the buffer!
			.word   0    # make sure we never overrun the buffer!
			.word   0    # make sure we never overrun the buffer!
			.word   0    # make sure we never overrun the buffer!
			.word   0    # make sure we never overrun the buffer!
.text
	# no matter what input they give us, we will modulo by 32, so that positions
	# "wrap around" over time.
	andi   $a0, $a0,0x1f        # this mask has 5 bits turned on!
	
	# read the address of the string
	la     $v0, getNextLetter_BUF
	
	# offset by the parameter
	add    $v0, $v0,$a0
	
	# end then read the character there, as the value we'll return.
	lb     $v0, 0($v0)
	
	# standard epilogue
	lw     $ra, 4($sp)
	lw     $fp, 0($sp)
	addiu  $sp, $sp, 24
	jr     $ra

.globl main
main: 
   
    # collatz_line(45)
	addi  $a0, $zero, 45
	jal   collatz_line
    
       
    # collatz_line(1)
	addi  $a0, $zero, 1
	jal   collatz_line
    
       
    # collatz_line(239)
	addi  $a0, $zero, 239
	jal   collatz_line
    
       
    # collatz_line(19)
	addi  $a0, $zero, 19
	jal   collatz_line
    
       
    # collatz_line(4)
	addi  $a0, $zero, 4
	jal   collatz_line
    
       
    # collatz_line(5)
	addi  $a0, $zero, 5
	jal   collatz_line
    
       
    # collatz_line(392)
	addi  $a0, $zero, 392
	jal   collatz_line
    
       
    # collatz_line(593)
	addi  $a0, $zero, 593
	jal   collatz_line
    
       
    # collatz_line(55)
	addi  $a0, $zero, 55
	jal   collatz_line
    
       
    # collatz_line(2)
	addi  $a0, $zero, 2
	jal   collatz_line
    
       
    # collatz_line(563)
	addi  $a0, $zero, 563
	jal   collatz_line
    
       
    # collatz_line(95)
	addi  $a0, $zero, 95
	jal   collatz_line
    
       
    # collatz_line(25)
	addi  $a0, $zero, 25
	jal   collatz_line
    
       
    # collatz(45)
	addi  $a0, $zero, 45
	jal   collatz
    
       
    # collatz(1)
	addi  $a0, $zero, 1
	jal   collatz
    
       
    # collatz(239)
	addi  $a0, $zero, 239
	jal   collatz
    
       
    # collatz(19)
	addi  $a0, $zero, 19
	jal   collatz
    
       
    # collatz(4)
	addi  $a0, $zero, 4
	jal   collatz
    
       
    # collatz(5)
	addi  $a0, $zero, 5
	jal   collatz
    
       
    # collatz(392)
	addi  $a0, $zero, 392
	jal   collatz
    
       
    # collatz(593)
	addi  $a0, $zero, 593
	jal   collatz
    
       
    # collatz(55)
	addi  $a0, $zero, 55
	jal   collatz
    
       
    # collatz(2)
	addi  $a0, $zero, 2
	jal   collatz
    
       
    # collatz(563)
	addi  $a0, $zero, 563
	jal   collatz
    
       
    # collatz(95)
	addi  $a0, $zero, 95
	jal   collatz
    
       
    # collatz(25)
	addi  $a0, $zero, 25
	jal   collatz
    
    
.data
SEARCH_STRING0:   .asciiz ""
SEARCH_STRING1:   .asciiz "%"
SEARCH_STRING2:   .asciiz "iot%"
SEARCH_STRING3:   .asciiz "i%"
SEARCH_STRING4:   .asciiz "woirj%zia%"
SEARCH_STRING5:   .asciiz "woer   %"

.text
	la      $a0, SEARCH_STRING0
	jal     percentSearch
	add     $t0, $v0,$zero
    
    addi    $v0, $zero,1              
	add     $a0, $t0,$zero
	syscall                 # print percentSearch(SEARCH_STRING0)
	
	addi    $v0, $zero,11             
	addi    $a0, $zero,0xa
	syscall                 # print newline
    
    	la      $a0, SEARCH_STRING1
	jal     percentSearch
	add     $t0, $v0,$zero
    
    addi    $v0, $zero,1              
	add     $a0, $t0,$zero
	syscall                 # print percentSearch(SEARCH_STRING1)
	
	addi    $v0, $zero,11             
	addi    $a0, $zero,0xa
	syscall                 # print newline
    
    	la      $a0, SEARCH_STRING2
	jal     percentSearch
	add     $t0, $v0,$zero
    
    addi    $v0, $zero,1              
	add     $a0, $t0,$zero
	syscall                 # print percentSearch(SEARCH_STRING2)
	
	addi    $v0, $zero,11             
	addi    $a0, $zero,0xa
	syscall                 # print newline
    
    	la      $a0, SEARCH_STRING3
	jal     percentSearch
	add     $t0, $v0,$zero
    
    addi    $v0, $zero,1              
	add     $a0, $t0,$zero
	syscall                 # print percentSearch(SEARCH_STRING3)
	
	addi    $v0, $zero,11             
	addi    $a0, $zero,0xa
	syscall                 # print newline
    
    	la      $a0, SEARCH_STRING4
	jal     percentSearch
	add     $t0, $v0,$zero
    
    addi    $v0, $zero,1              
	add     $a0, $t0,$zero
	syscall                 # print percentSearch(SEARCH_STRING4)
	
	addi    $v0, $zero,11             
	addi    $a0, $zero,0xa
	syscall                 # print newline
    
    	la      $a0, SEARCH_STRING5
	jal     percentSearch
	add     $t0, $v0,$zero
    
    addi    $v0, $zero,1              
	add     $a0, $t0,$zero
	syscall                 # print percentSearch(SEARCH_STRING5)
	
	addi    $v0, $zero,11             
	addi    $a0, $zero,0xa
	syscall                 # print newline
    
    addi    $a0, $zero,3               # letterTree(3)
	jal     letterTree                
	add     $t0, $v0,$zero             
	
	addi    $v0, $zero,1               
	add     $a0, $t0,$zero
	syscall                             # print letterTree(3)
	
	addi    $v0, $zero,11             
	addi    $a0, $zero,0xa
	syscall                             # print newline
    
    addi    $a0, $zero,25               # letterTree(25)
	jal     letterTree                
	add     $t0, $v0,$zero             
	
	addi    $v0, $zero,1               
	add     $a0, $t0,$zero
	syscall                             # print letterTree(25)
	
	addi    $v0, $zero,11             
	addi    $a0, $zero,0xa
	syscall                             # print newline
    
    addi    $a0, $zero,325               # letterTree(325)
	jal     letterTree                
	add     $t0, $v0,$zero             
	
	addi    $v0, $zero,1               
	add     $a0, $t0,$zero
	syscall                             # print letterTree(325)
	
	addi    $v0, $zero,11             
	addi    $a0, $zero,0xa
	syscall                             # print newline
    
    addi    $a0, $zero,2359               # letterTree(2359)
	jal     letterTree                
	add     $t0, $v0,$zero             
	
	addi    $v0, $zero,1               
	add     $a0, $t0,$zero
	syscall                             # print letterTree(2359)
	
	addi    $v0, $zero,11             
	addi    $a0, $zero,0xa
	syscall                             # print newline
    
    addi    $a0, $zero,395               # letterTree(395)
	jal     letterTree                
	add     $t0, $v0,$zero             
	
	addi    $v0, $zero,1               
	add     $a0, $t0,$zero
	syscall                             # print letterTree(395)
	
	addi    $v0, $zero,11             
	addi    $a0, $zero,0xa
	syscall                             # print newline
    
    addi    $a0, $zero,2954               # letterTree(2954)
	jal     letterTree                
	add     $t0, $v0,$zero             
	
	addi    $v0, $zero,1               
	add     $a0, $t0,$zero
	syscall                             # print letterTree(2954)
	
	addi    $v0, $zero,11             
	addi    $a0, $zero,0xa
	syscall                             # print newline
    
    addi    $a0, $zero,35               # letterTree(35)
	jal     letterTree                
	add     $t0, $v0,$zero             
	
	addi    $v0, $zero,1               
	add     $a0, $t0,$zero
	syscall                             # print letterTree(35)
	
	addi    $v0, $zero,11             
	addi    $a0, $zero,0xa
	syscall                             # print newline
    
    addi    $a0, $zero,31               # letterTree(31)
	jal     letterTree                
	add     $t0, $v0,$zero             
	
	addi    $v0, $zero,1               
	add     $a0, $t0,$zero
	syscall                             # print letterTree(31)
	
	addi    $v0, $zero,11             
	addi    $a0, $zero,0xa
	syscall                             # print newline
    
    addi    $a0, $zero,5               # letterTree(5)
	jal     letterTree                
	add     $t0, $v0,$zero             
	
	addi    $v0, $zero,1               
	add     $a0, $t0,$zero
	syscall                             # print letterTree(5)
	
	addi    $v0, $zero,11             
	addi    $a0, $zero,0xa
	syscall                             # print newline
    
    addi    $a0, $zero,2               # letterTree(2)
	jal     letterTree                
	add     $t0, $v0,$zero             
	
	addi    $v0, $zero,1               
	add     $a0, $t0,$zero
	syscall                             # print letterTree(2)
	
	addi    $v0, $zero,11             
	addi    $a0, $zero,0xa
	syscall                             # print newline
    
    addi    $a0, $zero,19               # letterTree(19)
	jal     letterTree                
	add     $t0, $v0,$zero             
	
	addi    $v0, $zero,1               
	add     $a0, $t0,$zero
	syscall                             # print letterTree(19)
	
	addi    $v0, $zero,11             
	addi    $a0, $zero,0xa
	syscall                             # print newline
    
    