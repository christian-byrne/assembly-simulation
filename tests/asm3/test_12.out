45
1
239
19
4 2 1
5
392 196 98 49
593
55
2 1
563
95
25
45
136 68 34 17
52 26 13
40 20 10 5
16 8 4 2 1
collatz(45) completed after 4 calls to collatz_line().

1
collatz(1) completed after 0 calls to collatz_line().

239
718 359
1078 539
1618 809
2428 1214 607
1822 911
2734 1367
4102 2051
6154 3077
9232 4616 2308 1154 577
1732 866 433
1300 650 325
976 488 244 122 61
184 92 46 23
70 35
106 53
160 80 40 20 10 5
16 8 4 2 1
collatz(239) completed after 17 calls to collatz_line().

19
58 29
88 44 22 11
34 17
52 26 13
40 20 10 5
16 8 4 2 1
collatz(19) completed after 6 calls to collatz_line().

4 2 1
collatz(4) completed after 0 calls to collatz_line().

5
16 8 4 2 1
collatz(5) completed after 1 calls to collatz_line().

392 196 98 49
148 74 37
112 56 28 14 7
22 11
34 17
52 26 13
40 20 10 5
16 8 4 2 1
collatz(392) completed after 7 calls to collatz_line().

593
1780 890 445
1336 668 334 167
502 251
754 377
1132 566 283
850 425
1276 638 319
958 479
1438 719
2158 1079
3238 1619
4858 2429
7288 3644 1822 911
2734 1367
4102 2051
6154 3077
9232 4616 2308 1154 577
1732 866 433
1300 650 325
976 488 244 122 61
184 92 46 23
70 35
106 53
160 80 40 20 10 5
16 8 4 2 1
collatz(593) completed after 25 calls to collatz_line().

55
166 83
250 125
376 188 94 47
142 71
214 107
322 161
484 242 121
364 182 91
274 137
412 206 103
310 155
466 233
700 350 175
526 263
790 395
1186 593
1780 890 445
1336 668 334 167
502 251
754 377
1132 566 283
850 425
1276 638 319
958 479
1438 719
2158 1079
3238 1619
4858 2429
7288 3644 1822 911
2734 1367
4102 2051
6154 3077
9232 4616 2308 1154 577
1732 866 433
1300 650 325
976 488 244 122 61
184 92 46 23
70 35
106 53
160 80 40 20 10 5
16 8 4 2 1
collatz(55) completed after 41 calls to collatz_line().

2 1
collatz(2) completed after 0 calls to collatz_line().

563
1690 845
2536 1268 634 317
952 476 238 119
358 179
538 269
808 404 202 101
304 152 76 38 19
58 29
88 44 22 11
34 17
52 26 13
40 20 10 5
16 8 4 2 1
collatz(563) completed after 13 calls to collatz_line().

95
286 143
430 215
646 323
970 485
1456 728 364 182 91
274 137
412 206 103
310 155
466 233
700 350 175
526 263
790 395
1186 593
1780 890 445
1336 668 334 167
502 251
754 377
1132 566 283
850 425
1276 638 319
958 479
1438 719
2158 1079
3238 1619
4858 2429
7288 3644 1822 911
2734 1367
4102 2051
6154 3077
9232 4616 2308 1154 577
1732 866 433
1300 650 325
976 488 244 122 61
184 92 46 23
70 35
106 53
160 80 40 20 10 5
16 8 4 2 1
collatz(95) completed after 38 calls to collatz_line().

25
76 38 19
58 29
88 44 22 11
34 17
52 26 13
40 20 10 5
16 8 4 2 1
collatz(25) completed after 7 calls to collatz_line().

-1
0
3
1
5
7
v
ii
hhh
wwww
ooooo
hhhhhh
eeeeeee
iiiiiiii
ttttttttt
27
v
tt
eee
tttt
jjjjj
125
v
ww
hhh
hhhh
aaaaa
tttttt
1950
v
ww
eee
wwww
9436
v
tt
ooo
aaaa
ooooo
wwwwww
ooooooo
wwwwwwww
ttttttttt
iiiiiiiiii
eeeeeeeeeee
tttttttttttt
jjjjjjjjjjjjj
hhhhhhhhhhhhhh
5530
v
hh
aaa
8862
v
ii
hhh
wwww
ooooo
hhhhhh
eeeeeee
iiiiiiii
ttttttttt
315
v
31
v
ww
hhh
hhhh
aaaaa
tttttt
30
v
oo
jjj
hhhh
eeeee
hhhhhh
ooooooo
eeeeeeee
ttttttttt
eeeeeeeeee
aaaaaaaaaaa
oooooooooooo
ttttttttttttt
26
v
oo
hhh
tttt
ooooo
95
